spec.html

```diff
@@ -36557,27 +36557,73 @@
       <p>The first use case is interoperable with the semantics of |Block| level function declarations provided by ECMAScript 2015. Any pre-existing ECMAScript code that employs that use case will operate using the Block level function declarations semantics defined by clauses 9, 13, and 14 of this specification.</p>
       <p>ECMAScript 2015 interoperability for the second and third use cases requires the following extensions to the clause <emu-xref href="#sec-ordinary-and-exotic-objects-behaviours"></emu-xref>, clause <emu-xref href="#sec-ecmascript-language-functions-and-classes"></emu-xref>, clause <emu-xref href="#sec-eval-x"></emu-xref> and clause <emu-xref href="#sec-globaldeclarationinstantiation"></emu-xref> semantics.</p>
       <p>If an ECMAScript implementation has a mechanism for reporting diagnostic warning messages, a warning should be produced when code contains a |FunctionDeclaration| for which these compatibility semantics are applied and introduce observable differences from non-compatibility semantics. For example, if a var binding is not introduced because its introduction would create an early error, a warning message should not be produced.</p>
+      <emu-annex id="sec-web-compat-block-static-semantics-lexicallydeclarednames">
+        <h1>Changes to Block LexicallyDeclaredNames</h1>
+        <p>In Static Semantics: LexicallyDeclaredNames of <emu-grammar>StatementList : StatementList StatementListItem</emu-grammar> (<emu-xref href="#sec-block-static-semantics-lexicallydeclarednames"></emu-xref>) the following is performed in place of step 2:</p>
+        <emu-alg>
+          1. If the production is contained in strict mode code, then
+            1. Append to _names_ the elements of the TopLevelLexicallyDeclaredNames of |StatementListItem|.
+          1. Else,
+            1. If |StatementListItem| is <emu-grammar>StatementListItem : Declaration</emu-grammar> and |Declaration| is <emu-grammar>Declaration : HoistableDeclaration</emu-grammar> and |HoistableDeclaration| is <emu-grammar>HoistableDeclaration : FunctionDeclaration</emu-grammar> , then
+              1. Let _fn_ be the sole element of the BoundNames of |FunctionDeclaration|.
+              1. If _names_ does not include the element _fn_, append _fn_ to _names_.
+            1. Else,
+              1. Append to _names_ the elements of the TopLevelLexicallyDeclaredNames of |StatementListItem|.
+        </emu-alg>
+      <emu-note>
+        Multiple |FunctionDeclaration|s for the same name inside a block is allowed when not in strict mode code.
```

Comment:
"inside a block" -> "inside a |Block| or |CaseClause|". Also this makes me wonder how duplicate function declarations in switch-statements should be handled when the duplicates don't appear in the same case-clause:
```js
switch (v) {
  case 0: function f() {}
  case 1: function f() {}
}
```

If this also needs to be supported, it's probably easier to change the early error conditions from:
> It is a Syntax Error if the LexicallyDeclaredNames of StatementList contains any duplicate entries.

to:
> It is a Syntax Error if the LexicallyDeclaredNames of StatementList contains any duplicate entries,
unless the source code matching this production is not strict code and the duplicate entries are only bound by FunctionDeclarations.